### ISSUE TITLE
PyTorch Lightning and better T-S distillation

### ISSUE THREAD

**Issue by DiTo97** [DiTo97/Rf-PIQA#3]
I'm trying to model the following problem with neural networks using knowledge distillation. I will lay out the setting. Your goal is to suggest the best possible options I could opt for. I am building a framework around all of that, and would love users to have all the best options at their disposal to choose from.

So, we have two NNs, a teacher and a student. Roughly speaking, the two models take images as input as produce a single point estimate score. The teacher is not only larger, but also accept one extra modality than the student. The distillation goal would be to train an equally accurate model while dropping that extra modality in the student. The models are for image quality assessment (IQA). Precisely, for panorama IQA. To dive deeper into the inputs format, the teacher model receives records made of the large high-resolution panorama image, and the corresponding set of lower resolution constituent images. A large backbone processes the panorama image into a panorama feature map, while a smaller backbone processes each constituent image, pooling their representation together into a single constituent feature map. The diff feature map between the two is the overall network feature map passed to the model head. On the contrary, the student model receives records made only of the panorama image, only has the panorama backbone, and the overall network feature map coincides with the panorama feature map. The options for the head are basically identically in both cases, starting from the overall feature map. The end user can choose between two different types of heads: a point estimate head, or a prediction intervals head (PIVEN). The point estimate head receives the overall feature map and produces a single floating point quality score, between 0 and 1 (or 0 and 100, whatever scale you think gives the model more stability during training). On the other end, the PIVEN head receives the overall feature map, and produces a feature vector of size three, where the first two entries map the upper and lower bound of the prediction interval, and the third and last entry is a normalized weight (passed through the sigmoid function) between 0 and 1 that polarizes where the value is within the prediction interval. The point estimate from the PIVEN head can be recovered with a specific formula that uses the normalized weight to identify the value estimate within the prediction interval, with the added bonus that the size of the interval can be used as a noise, and therefore confidence, estimate. The point estimates are evaluated against the target quality score using a standard MSE loss (or an equivalent regression loss).
The framework should allow end users to create any of the six combinations of models (teacher + point head, teacher + PIVEN head, student + point head, student + PIVEN head, student + point head via distillation, student + PIVEN head via distillation), with any regression loss between point estimates and target quality scores. You have to tell me how would you structure knowledge distillation, where would you tie it, and if there are multiple valuable options for it that we should offer to our end users. Remember, I don't need neither generic distillation options, nor all possibile knowledge distillation options for the case I just described, rather only the valuable ones that are achievable and would make sense to offer.

**Comment by DiTo97** [comment]
To address your problem, here's a structured approach for implementing knowledge distillation (KD) in your framework, focusing on the most valuable and achievable options:

---

### **1. Core Distillation Strategies**
These strategies are tailored to your specific setup, where the student lacks the teacher's extra modality (constituent images).

#### **a. Feature-Level Distillation**
- **Objective**: Align the student's panorama features with the teacher’s *diff feature map* (panorama features − pooled constituent features).
- **Implementation**:
  - Add a **projection head** (e.g., 1x1 convolution or linear layer) to the student’s panorama backbone to match the dimensionality of the teacher’s diff features.
  - Use **MSE loss** or **cosine similarity loss** between the projected student features and the teacher’s diff features.
- **Why**: Forces the student to mimic the teacher’s fused representation of panorama and constituent images, compensating for missing modality.

#### **b. Output-Level Distillation**
- **Objective**: Match the student’s predictions to the teacher’s outputs.
  - **For Point Heads**: 
    - Train the student with a **joint loss**: `α * MSE(student, ground_truth) + β * MSE(student, teacher)`.
  - **For PIVEN Heads**:
    - Match interval bounds (lower/upper) with MSE and the weight with **KL divergence** or **BCE** (since it’s sigmoid-normalized).
    - Optionally include the teacher’s derived point estimate (via PIVEN formula) in the loss.
- **Why**: Directly transfers the teacher’s predictive behavior, including uncertainty estimates (for PIVEN).

#### **c. Hybrid Distillation**
- Combine feature-level and output-level losses:
  ```
  Total Loss = α * Regression Loss (student vs. GT) 
               + β * Feature Distillation Loss 
               + γ * Output Distillation Loss
  ```
- **Why**: Balances direct prediction alignment with internal representation matching.

---

### **2. Framework Design for User Flexibility**
Allow users to choose combinations of distillation strategies and losses:

#### **a. Distillation Modes**
1. **Feature Distillation Only**: For users prioritizing representation learning.
2. **Output Distillation Only**: Simpler setup, ideal for quick experimentation.
3. **Hybrid Mode**: Best of both worlds, recommended for maximal accuracy.

#### **b. Loss Options**
- **Regression Losses**:
  - MSE, MAE, Huber (user-selectable for both teacher and student).
- **Feature Distillation Losses**:
  - MSE, Cosine Similarity, or Projection-based (e.g., Wasserstein distance).
- **PIVEN-Specific Losses**:
  - Interval loss (MSE for bounds), Weight loss (KL/BCE), or combined PIVEN loss.

#### **c. Weighting Schemes**
- Allow users to set weights (`α, β, γ`) for loss components.
- Provide annealing schedules (e.g., decay `β` over epochs to phase out feature distillation).

---

### **3. PIVEN-Specific Enhancements**
- **Teacher Confidence Transfer**:
  - If the teacher uses a PIVEN head, use its interval width (upper − lower) as a **confidence signal** for the student. For example:
    - Add a loss term to minimize the difference between teacher and student interval widths.
- **Point Estimate Recovery**:
  - For distillation to a student with a point head, derive the teacher’s point estimate from its PIVEN head and include it in the output loss.

---

### **4. Implementation Roadmap**
1. **Pre-Trained Teacher**: Freeze the teacher during distillation.
2. **Feature Alignment**:
   - Add a configurable projection layer to the student (enabled only for feature distillation).
3. **Loss Modularity**:
   - Implement loss components as pluggable modules (users can enable/disable them).
4. **PIVEN Utilities**:
   - Include functions to derive point estimates from PIVEN outputs and compute interval metrics.

---

### **5. Recommended Defaults**
- **For Point Heads**: Hybrid distillation (feature + output) with `α=0.7, β=0.3`.
- **For PIVEN Heads**: Output distillation on intervals + weights, with feature distillation for high-stakes tasks.
- **Loss Weights**: Start with equal weights (`α=β=0.5`) and let users tune.

---

### **6. Key User Options**
| **Component**       | **Options**                                                                 |
|----------------------|-----------------------------------------------------------------------------|
| Distillation Mode    | Feature, Output, Hybrid                                                    |
| Feature Loss         | MSE, Cosine, Projection + MSE                                              |
| Output Loss (Point)  | MSE, MAE, Huber                                                            |
| Output Loss (PIVEN)  | Interval MSE + Weight KL, Interval MAE + Weight BCE                        |
| Loss Weighting       | Fixed weights, Annealed weights (user-defined schedule)                    |
| Teacher Head Source  | Use teacher’s point head or PIVEN-derived point estimate for distillation  |

---

### **Why This Works**
- **Feature Distillation** bridges the modality gap by aligning the student’s panorama features with the teacher’s fused representation.
- **Output Distillation** ensures the student matches the teacher’s predictions, including uncertainty (for PIVEN).
- **Hybrid Mode** maximizes accuracy by combining both strategies, mimicking human learning (concepts + results).

This structure balances flexibility and performance, letting users prioritize simplicity (output-only) or depth (hybrid).

**Comment by DiTo97** [comment]
Here's a PyTorch Lightning implementation with configurable knowledge distillation options. I'll focus on the critical parts of the training loop and loss handling:

```python
from dataclasses import dataclass
import torch
import torch.nn as nn
import torch.nn.functional as F
import pytorch_lightning as pl

# Configuration Dataclass
@dataclass
class DistillConfig:
    head_type: str = 'point'  # 'point' or 'piven'
    distill_mode: str = 'hybrid'  # 'feature', 'output', 'hybrid'
    use_feature_projection: bool = True
    feature_loss_weight: float = 0.5
    output_loss_weight: float = 0.5
    anneal_feature_weight: bool = False
    anneal_epochs: int = 10

# Feature Projection Module
class FeatureProjection(nn.Module):
    def __init__(self, student_feat_dim, teacher_feat_dim):
        super().__init__()
        self.projection = nn.Sequential(
            nn.Conv2d(student_feat_dim, teacher_feat_dim, 1),
            nn.ReLU(inplace=True)
        )
    
    def forward(self, x):
        return self.projection(x)

# Lightning Module
class IQADistillation(pl.LightningModule):
    def __init__(self, teacher, student, config: DistillConfig):
        super().__init__()
        self.teacher = teacher
        self.student = student
        self.config = config
        
        # Freeze teacher
        for param in self.teacher.parameters():
            param.requires_grad = False
            
        # Initialize feature projection if needed
        if self.config.distill_mode in ['feature', 'hybrid']:
            self.feature_projector = FeatureProjection(
                student_feat_dim=student.backbone.out_channels,
                teacher_feat_dim=teacher.backbone.out_channels
            )
            
        # Losses
        self.mse_loss = nn.MSELoss()
        self.piven_loss = self._piven_loss  # Custom implementation needed

    def forward(self, panos, constituents=None):
        # Used only for inference
        return self.student(panos)

    def _get_losses(self, student_out, teacher_out, features):
        losses = {}
        
        # Main task loss
        if self.config.head_type == 'point':
            losses['main'] = self.mse_loss(student_out, features['target'])
        else:
            losses['main'] = self.piven_loss(student_out, features['target'])

        # Feature distillation loss
        if self.config.distill_mode in ['feature', 'hybrid']:
            proj_student_feat = self.feature_projector(features['student'])
            losses['feature'] = self.mse_loss(proj_student_feat, features['teacher_diff'])

        # Output distillation loss
        if self.config.distill_mode in ['output', 'hybrid']:
            if self.config.head_type == 'point':
                losses['output'] = self.mse_loss(student_out, teacher_out)
            else:
                # PIVEN-specific distillation
                losses['interval'] = self.mse_loss(student_out[:,:2], teacher_out[:,:2])
                losses['weight'] = F.binary_cross_entropy(student_out[:,2], teacher_out[:,2])

        return losses

    def training_step(self, batch, batch_idx):
        panos, constituents, targets = batch
        teacher_feat = self.teacher.backbone(panos, constituents)
        
        # Get features
        with torch.no_grad():
            teacher_diff = self.teacher.get_diff_features(panos, constituents)
            teacher_out = self.teacher.head(teacher_diff)

        student_feat = self.student.backbone(panos)
        student_out = self.student.head(student_feat)

        # Calculate losses
        features = {
            'student': student_feat,
            'teacher_diff': teacher_diff,
            'target': targets
        }
        losses = self._get_losses(student_out, teacher_out, features)

        # Combine losses with weights
        total_loss = losses['main']
        if 'feature' in losses:
            total_loss += self.config.feature_loss_weight * losses['feature']
        if 'output' in losses:
            total_loss += self.config.output_loss_weight * losses['output']
        if 'interval' in losses:  # PIVEN case
            total_loss += 0.5*(losses['interval'] + losses['weight'])

        # Logging
        self.log_dict({f'train_{k}': v for k,v in losses.items()})
        return total_loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.student.parameters())

    def on_train_epoch_start(self):
        # Annealing logic
        if self.config.anneal_feature_weight:
            progress = self.current_epoch / self.config.anneal_epochs
            self.config.feature_loss_weight *= (1 - progress)

    # PIVEN Loss Implementation (example)
    def _piven_loss(self, pred, target):
        lower, upper, weight = pred[:,0], pred[:,1], pred[:,2]
        point_pred = weight * lower + (1 - weight) * upper
        return self.mse_loss(point_pred, target) + 0.5*self.mse_loss(upper - lower, torch.zeros_like(upper))
```

Key implementation notes:

1. **Architecture Assumptions**:
```python
# Teacher forward logic
class Teacher(nn.Module):
    def get_diff_features(self, panos, constituents):
        pano_feat = self.pano_backbone(panos)
        const_feat = self.const_backbone(constituents)
        return pano_feat - const_feat.mean(dim=0)  # Example pooling

# Student forward logic
class Student(nn.Module):
    def forward(self, panos):
        feat = self.backbone(panos)
        return self.head(feat)
```

2. **Configuration Options**:
```python
# Example configurations
point_head_config = DistillConfig(
    head_type='point',
    distill_mode='hybrid',
    use_feature_projection=True
)

piven_config = DistillConfig(
    head_type='piven',
    distill_mode='output',
    anneal_feature_weight=True,
    anneal_epochs=20
)
```

3. **Training Flow**:
- Teacher is always frozen
- Feature projection only used when doing feature distillation
- Automatic mixed precision handled by Lightning
- Loss weights annealed linearly if configured

4. **Critical Distillation Paths**:
- **Feature Distillation**: 
  ```text
  Student Backbone → Projection → MSE → Teacher Diff Features
  ```
- **Output Distillation**:
  ```text
  Student Head Output → MSE/KL → Teacher Head Output
  ```
- **Hybrid Mode**: Sum of both losses

5. **PIVEN-Specific Handling**:
- Special loss calculation for interval bounds and weights
- Point estimate reconstruction handled internally
- Confidence estimation from interval width

To use this implementation:

```python
teacher = Teacher()  # Pretrained
student = Student()
config = DistillConfig(distill_mode='hybrid')

model = IQADistillation(teacher, student, config)
trainer = pl.Trainer(max_epochs=50)
trainer.fit(model, datamodule)
```

This implementation provides:
- Flexible distillation mode switching
- Automatic feature projection matching
- Configurable loss weighting
- PIVEN-specific handling
- Loss annealing capabilities
- Full logging integration

The hardest parts are handled by:
1. Feature projection matching between different backbone architectures
2. Combined loss calculation with multiple modalities
3. PIVEN-specific output interpretation
4. Weight annealing integrated into the training loop
